### DISCUSSION TITLE
UW Release Requirements

### DISCUSSION THREAD

**Discussion by christinaholtNOAA** [ufs-community/uwtools#164] [Q&A]
Over the years, we've gathered (mostly verbally) the requirements for the Unified Workflow standalone tools. Mostly these are about what is needed for the tool to be adopted in the UFS Apps/components.

- The UW packages should meet, at minimum, existing requirements/tests as implemented in the host App/component. 
- Each tool should be a separate entity that handles templating and creating a configuration file.
- There should be some way to warn the user or fail when a known invalid entry is provided. 
- Supported conda environments should be installed on all the standard platforms before UW can be adopted by a host App/component.
- Inclusion of the UW Tools will be through the relevant externals management system of the host App/component -- git submodules or manage externals.
- Installation of the package needs to play nicely with both conda and virtual env users, and to know how to install all package dependencies.
   - Include both an  environment.yaml (conda) and a requirements.txt (venv) 
- For the package to be supported on standard platforms, it will need to be "installable" locally and lean on a centrally installed conda environment to handle its external 3rd party dependencies.
- Test suite should capture any of the existing test cases included in host App/component.

Specifics of the config management tools for the first release:
- We should have tools that allow us to treat _any_ config file as a flat file template.
- We need a way to reference other keys, environment variables, and to do math and structure control.
- We need to be able to overwrite existing values in a prepared dictable config file.

Which ones have I missed here? Other than requests for demos, what additionsl outstanding questions do we still have about this release. Others may come out of follow-up demos.

**Comment by frolovsa** [comment]
@christinaholtNOAA I think you have it under "We should have tools that allow us to treat any config file as a flat file template." but i thought i will clarify it here. 

Working with JCSDA and EMC, it becoming clear that "centers" will be responsible for a reference template configuration. The format of reference centers might be different. So a successful tool should be able to recognize multiple templating languages. 

It will be good to understand how UW config management solution will be similar/different/complimentary/orthogonal from what is already developed and used for GDASapp and EWOK.
